### PR TITLE
Core: HadoopTable needs to skip file cleanup after task failure under some boundary conditions.

### PR DESCRIPTION
**Pull request by BsoBird**
We found that under some boundary conditions, HadoopCatalogTable will suffer from data file loss.
Since our Spark uses an LLAP-like runtime model(Long live and process), we need to take into account some additional runtime exceptions.
The problem is in BaseFileRewriteAction::doReplace.
When an exception is caught, if it is an unrecognised exception, Spark will clean up the data file for this commit.(only datafile,not have metafile)
But if we use HadoopCatalog, its commit method is not atomic.
HadoopCatalog's commit method needs to operate the file in hdfs several times, if in the process of operating the file, due to other DAG triggered OOM, and it happens that Iceberg's metadata has been submitted, but the entire commit method is not finished. The call to the commit method will throw a RunTimeException. Caught by the doReplace method. Triggers a cleanup of the data file.
The end result is that the metafile is written, but the corresponding datafile is cleaned up. 

I've changed the logic of commit so that when the versionhint is modified, if there is an exception at that point, we will skip the data cleanup session.

FIX: https://github.com/apache/iceberg/issues/9327

### CODE REVIEW

**Comment by BsoBird** [comment]
@RussellSpitzer  Hi. can you check this?

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 1]
```diff
@@ -129,7 +130,7 @@ public TableMetadata refresh() {
 
   @Override
   public void commit(TableMetadata base, TableMetadata metadata) {
-    Pair<Integer, TableMetadata> current = versionAndMetadata();
+    Pair<Long, TableMetadata> current = versionAndMetadata();
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:133]: The change from int to long seems unrelated?

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java [hunk 2]
```diff
@@ -146,29 +147,37 @@ public void commit(TableMetadata base, TableMetadata metadata) {
...
-    writeVersionHint(nextVersion);
+      // update the best-effort version pointer
+      versionCommitSuccess = writeVersionHint(nextVersion);
 
-    deleteRemovedMetadataFiles(base, metadata);
+      deleteRemovedMetadataFiles(base, metadata);
 
-    this.shouldRefresh = true;
+      this.shouldRefresh = true;
+    }catch (Throwable e){
```
Comment (RussellSpitzer) [inline core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:176]: This is a bit over cautious, only after line 166 do we believe the commit state is unknown, prior to this we know it was a faiulre

File: core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java (hunk 1 above)
Comment (BsoBird) [reply core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java:133]: Yes, I just thought that versionId defined as integer type is easy to overflow, so I changed it to long type.

**Comment by BsoBird** [comment]
@RussellSpitzer  I have a question, since the versionId we defined is of type Integer, it looks like it could easily overflow. Isn't there some problem when it overflows?

**Comment by BsoBird** [comment]
@RussellSpitzer 
Hi. I've found that if we want to be able to simulate exceptions better, we need to use PowerMock to proxy some private methods. There are some catalogs where the execution of the commit method is not atomic. Simply using mokito is not a good way to simulate unexpected situations. Can I import PowerMock?
